applications/FluidDynamicsApplication/custom_elements/fluid_adjoint_element.cpp

```diff
@@ -0,0 +1,717 @@
+        AddFluidShapeDerivatives(rOutput, rCurrentProcessInfo);
+    } else {
+        KRATOS_ERROR << "Sensitivity variable " << rSensitivityVariable
+                     << " not supported." << std::endl;
+    }
+
+    KRATOS_CATCH("")
+}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+void FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::Calculate(
+    const Variable<Vector>& rVariable,
+    Vector& rOutput,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    if (rVariable == PRIMAL_RELAXED_SECOND_DERIVATIVE_VALUES) {
+        if (rOutput.size() != TElementLocalSize) {
+            rOutput.resize(TElementLocalSize);
+        }
+
+        IndexType local_index = 0;
+        for (IndexType i_node = 0; i_node < TNumNodes; ++i_node) {
+            const auto& r_node = this->GetGeometry()[i_node];
+            rOutput[local_index++] = r_node.GetValue(RELAXED_ACCELERATION_X);
+            rOutput[local_index++] = r_node.GetValue(RELAXED_ACCELERATION_Y);
+
+            if (TDim == 3)
+                rOutput[local_index++] = r_node.GetValue(RELAXED_ACCELERATION_Z);
+
+            rOutput[local_index++] = 0.0;
+        }
+    } else {
+        KRATOS_ERROR << "Unsupported variable requested Calculate method. [ rVariable.Name() = " << rVariable.Name() << " ].\n";
+    }
+
+    KRATOS_CATCH("");
+}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+std::string FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::Info() const
+{
+    std::stringstream buffer;
+    buffer << "FluidAdjointElement #" << Element::Id();
+    return buffer.str();
+}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+void FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::PrintInfo(std::ostream& rOStream) const
+{
+    rOStream << "FluidAdjointElement #" << Element::Id();
+}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+void FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::PrintData(std::ostream& rOStream) const
+{
+    Element::pGetGeometry()->PrintData(rOStream);
+}
+
+///@}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+void FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::AddFluidResidualsContributions(
+    VectorType& rOutput,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    const auto& integration_method = TAdjointElementData::TResidualsDerivatives::GetIntegrationMethod();
+
+    Vector Ws;
+    Matrix Ns;
+    ShapeFunctionDerivativesArrayType dNdXs;
+    this->CalculateGeometryData(Ws, Ns, dNdXs, integration_method);
+
+    typename TAdjointElementData::Primal::Data element_data(*this, *mpConstitutiveLaw, rCurrentProcessInfo);
+    typename TAdjointElementData::Primal::ResidualsContributions residual_contributions(element_data);
+
+    VectorF residual = ZeroVector(TElementLocalSize);
```

Comment:
Where is this type defined? I've never seen it :open_mouth:

Reply:
Its here
https://github.com/KratosMultiphysics/Kratos/blob/c44195d584f01124791e43c2184d8e0a9e6ce61a/applications/FluidDynamicsApplication/custom_elements/fluid_adjoint_element.h#L92